### PR TITLE
Added support for routing parameters.

### PR DESCRIPTION
**Pull request by cdahlqvist**
I have added support for routing when indexing documents.

### CODE REVIEW

**Comment by jordansissel** [comment]
Code looks good, I think. I haven't tested it.

**Comment by cdahlqvist** [comment]
I copied the parameter description from PR #64 to my code as it is better and more descriptive.

**Comment by masteinhauser** [comment]
Ah, I didn't realize there was more work which was needed. Though, I definitely missed adding a test! I'll close out #64 in support of this.

**Comment by masteinhauser** [comment]
Is there anything else needed for this PR to be reviewed, tested and merged?

It looks like the tests failed due to a CLA issue but it also seems like the QA process is not updated for the new plugins repo organization.

**Comment by jordansissel** [comment]
@masteinhauser time. We'll tend to this soon. We are focusing on other priorities at this time.

**Comment by masteinhauser** [comment]
Ah, understood. I will package and keep my own internally. Thanks!

File: spec/outputs/elasticsearch_spec.rb [hunk 1]
```diff
@@ -73,6 +73,75 @@
     end
   end
 
+  describe "ship lots of events w/ default index_type and fixed routing key", :elasticsearch => true do
+    # Generate a random index name
+    index = 10.times.collect { rand(10).to_s }.join("")
```
Comment (purbon) [inline spec/outputs/elasticsearch_spec.rb]: This should go into a let variable (more info at https://relishapp.com/rspec/rspec-core/v/3-2/docs/helper-methods/let-and-let )

File: spec/outputs/elasticsearch_spec.rb [hunk 2]
```diff
@@ -73,6 +73,75 @@
...
+      ftw.post!("http://localhost:9200/#{index}/_refresh")
+
+      # Wait until all events are available.
+      Stud::try(10.times) do
+        data = ""
+        response = ftw.get!("http://127.0.0.1:9200/#{index}/_count?q=*")
+        response.read_body { |chunk| data << chunk }
+        result = LogStash::Json.load(data)
+        count = result["count"]
+        insist { count } == event_count
```
Comment (purbon) [inline spec/outputs/elasticsearch_spec.rb]: You should avoid using insist and use expect as we use now proper rspec syntax.

File: spec/outputs/elasticsearch_spec.rb [hunk 3]
```diff
@@ -73,6 +73,75 @@
...
+      response = ftw.get!("http://127.0.0.1:9200/#{index}/_search?q=*&size=1000&?routing=test")
+      data = ""
+      response.read_body { |chunk| data << chunk }
+      result = LogStash::Json.load(data)
+      count = result["count"]
+      insist { count } == event_count
+      result["hits"]["hits"].each do |doc|
+        # With no 'index_type' set, the document type should be the type
+        # set on the input
+        insist { doc["_type"] } == type
```
Comment (purbon) [inline spec/outputs/elasticsearch_spec.rb]: Is a good practice to break test into test examples with a single assertion on them, they use to be more clear and clever for everyone.

File: lib/logstash/outputs/elasticsearch.rb [hunk 4]
```diff
@@ -81,6 +81,10 @@ class LogStash::Outputs::ElasticSearch < LogStash::Outputs::Base
   # Elasticsearch with the same ID.
   config :document_id, :validate => :string, :default => nil
 
+  # A routing override to be applied to all processed events.
+  # This can be dynamic using the `%{foo}` syntax.
+  config :routing, :validate => :string, :default => nil
```
Comment (purbon) [inline lib/logstash/outputs/elasticsearch.rb]: This config variable should not have a default. But I agree is not only something for this, as it is already done with document id.

Defaults should be viable values.

Comment (cdahlqvist) [inline lib/logstash/outputs/elasticsearch.rb]: @purbon I have addressed this.

File: spec/outputs/elasticsearch_spec.rb (hunk 2 above)
Comment (cdahlqvist) [inline spec/outputs/elasticsearch_spec.rb]: @purbon I have modified this across the module.

**Comment by jordansissel** [comment]
@talevy can you take over review on this? @purbon is going on vacation (WOO) and I'd like to help move this forward for our friend @cdahlqvist :)

**Comment by cdahlqvist** [comment]
@talevy Added test for dynamic routing parameter as discussed.

**Comment by talevy** [comment]
@cdahlqvist thanks!

I'll try to check it out and test locally soon and get back to you!

**Comment by talevy** [comment]
@cdahlqvist why did you change all the `insist` to `expect`. can we leave that alone for now? Plus, I cannot run tests as is with that

**Comment by talevy** [comment]
@cdahlqvist I also see that there are a few `result['count']` retrievals for verifying number of responses. This field is only found in `_count` requests and not `_search`. to verify count of a search response you may want to look at `result['hits']['hits'].size`.

File: spec/outputs/elasticsearch_spec.rb [hunk 5]
```diff
@@ -56,7 +56,7 @@
         response.read_body { |chunk| data << chunk }
         result = LogStash::Json.load(data)
         count = result["count"]
-        insist { count } == event_count
+        expect { count } == event_count
```
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: we have traditionally been using `insist` for such assertions. I think we should keep it with `insist` for the sake of keeping this PR's main goal to add the `_routing` functionality

File: spec/outputs/elasticsearch_spec.rb [hunk 6]
```diff
@@ -66,13 +66,293 @@
...
+      expect { count } == event_count
+      result["hits"]["hits"].each do |doc|
+        # With no 'index_type' set, the document type should be the type
+        # set on the input
+        expect { doc["_type"] } == type
+        expect { doc["_index"] } == index
+        expect { doc["_source"]["message"] } == "hello world"
+      end
+
+      response = ftw.get!("http://127.0.0.1:9200/#{index}/_search?q=*&size=1000&?routing=not_test")
```
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: `&?routing=not_test` should not contain the `?`

I went ahead and removed this and test was successful

File: spec/outputs/elasticsearch_spec.rb [hunk 7]
```diff
@@ -66,13 +66,293 @@
...
+      Stud::try(10.times) do
+        data = ""
+        response = ftw.get!("http://127.0.0.1:9200/#{index}/_count?q=*")
+        response.read_body { |chunk| data << chunk }
+        result = LogStash::Json.load(data)
+        count = result["count"]
+        expect { count } == event_count
+      end
+
+      response = ftw.get!("http://127.0.0.1:9200/#{index}/_search?q=*&size=1000&?routing=test")
```
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: same as below, `&?routing` should be `&routing` at the end of that query

File: spec/outputs/elasticsearch_spec.rb [hunk 8]
```diff
@@ -66,13 +66,293 @@
...
```
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: no `?` before `routing`

File: spec/outputs/elasticsearch_spec.rb [hunk 9]
```diff
@@ -66,13 +66,293 @@
...
+      expect { count } == event_count
+      result["hits"]["hits"].each do |doc|
+        # With no 'index_type' set, the document type should be the type
+        # set on the input
+        expect { doc["_type"] } == type
+        expect { doc["_index"] } == index
+        expect { doc["_source"]["message"] } == "test"
+      end
+
+      response = ftw.get!("http://127.0.0.1:9200/#{index}/_search?q=*&size=1000&?routing=not_test")
```
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: no `?` before `routing`

File: spec/outputs/elasticsearch_spec.rb [hunk 10]
```diff
@@ -66,13 +66,293 @@
...
+      data = ""
+      response.read_body { |chunk| data << chunk }
+      result = LogStash::Json.load(data)
+      count = result["count"]
```
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: `result["count"]` should be `result['hits']['hits'].size`

File: spec/outputs/elasticsearch_spec.rb [hunk 11]
```diff
@@ -66,13 +66,293 @@
...
+        result = LogStash::Json.load(data)
+        count = result["count"]
+        expect { count } == event_count
+      end
+
+      response = ftw.get!("http://127.0.0.1:9200/#{index}/_search?q=*&size=1000&?routing=test")
+      data = ""
+      response.read_body { |chunk| data << chunk }
+      result = LogStash::Json.load(data)
+      count = result["count"]
```
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: `result["count"]` should be `result['hits']['hits'].size`

File: spec/outputs/elasticsearch_spec.rb [hunk 12]
```diff
@@ -66,13 +66,293 @@
...
+      Stud::try(10.times) do
+        data = ""
+        response = ftw.get!("http://127.0.0.1:9200/#{index}/_count?q=*")
+        response.read_body { |chunk| data << chunk }
+        result = LogStash::Json.load(data)
+        count = result["count"]
+        expect { count } == event_count
+      end
+
+      response = ftw.get!("http://127.0.0.1:9200/#{index}/_search?q=*&size=1000&?routing=test")
```
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: no `?` before `routing`

File: spec/outputs/elasticsearch_spec.rb [hunk 13]
```diff
@@ -66,13 +66,293 @@
...
+      expect { count } == event_count
+      result["hits"]["hits"].each do |doc|
+        # With no 'index_type' set, the document type should be the type
+        # set on the input
+        expect { doc["_type"] } == type
+        expect { doc["_index"] } == index
+        expect { doc["_source"]["message"] } == "hello world"
+      end
+
+      response = ftw.get!("http://127.0.0.1:9200/#{index}/_search?q=*&size=1000&?routing=not_test")
```
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: no `?` before `routing`

**Comment by talevy** [comment]
I also notice that many of the tests for routing simply check document counts from request with routing parameters. you may simply want to call the count api, and check count, instead of calling search

**Comment by cdahlqvist** [comment]
That is a very good point. Will modify to instead use the count API and correct the syntax.

**Comment by cdahlqvist** [comment]
@talevy Have cleaned up the routing tests and made them instead use the _count API. Did however not change back to using 'insist' as @purbon preferred the use of 'expect'.

**Comment by talevy** [comment]
we can go ahead and change `insist` to `expect` in a different PR, if you do not mind. Plus, I cannot seem to be able to run the tests with `expect`

**Comment by cdahlqvist** [comment]
@talevy Changed back to using expect. Hopefully it will work properly now.

**Comment by cdahlqvist** [comment]
@talevy Have merged changes from master.

File: spec/outputs/elasticsearch_spec.rb (hunk 2 above)
Comment (talevy) [inline spec/outputs/elasticsearch_spec.rb]: @purbon For the sake of this PR, I told @cdahlqvist we can address this at a later date in its own PR.

**Comment by talevy** [comment]
@cdahlqvist looks good! but, I am having trouble having the whole test-suite pass... your tests pass. other tests pass. but when I run it all together, things fail. I am trying to debug it. once that is resolved, I will merge this in!

**Comment by talevy** [comment]
so `master` is ready. but I cant seem to get your tests to pass. they conflict and cause this:

```
Apr 03, 2015 1:04:35 PM org.apache.http.impl.execchain.RetryExec execute
INFO: I/O exception (java.net.SocketException) caught when processing request to {}->http://127.0.0.1:9200: Bad file descriptor

```

**Comment by talevy** [comment]
https://github.com/cdahlqvist/logstash-output-elasticsearch/blob/master/spec/outputs/elasticsearch_spec.rb#L701-L704

in the lines above. the mocked actions need to be updated to include the `:_routing` param.

for example:

```
let(:action1) { ["index", {:_id=>nil, :_index=>"logstash-2014.11.17", :_type=>"logs"}, event1] }
```
should be updated to:
```
let(:action1) { ["index", {:_id=>nil, :_routing=>nil, :_index=>"logstash-2014.11.17", :_type=>"logs"}, event1] }
```

same goes for `:action2`

**Comment by iemem15** [comment]
Hi, When is going to be available to use it?

**Comment by cdahlqvist** [comment]
@talevy Have added the routing parameter.

**Comment by talevy** [comment]
cool! tests are looking good @cdahlqvist!

can you go ahead and `git pull` and `git rebase master` so we can merge this in?

**Comment by cdahlqvist** [comment]
@talevy Done.

**Comment by talevy** [comment]
Thank you!

LGTM

**Comment by crickeys** [comment]
Does this currently work for UPDATES in the downloadable or YUM versions of logstash? I'm using logstash-1.4.2-1_2c0f5a1.noarch

I'm having issues with updating parent/child documents. I can't seem to add any routing information when doing an UPDATE action. I keep getting  :message=>"Unknown setting 'routing' for elasticsearch", :level=>:error}

I've also tried logstash-1.5.0.rc2-1.noarch and same message. I'm new to logstash/elasticsearch so forgive me if this is obvious.

**Comment by talevy** [comment]
@crickeys we are still focusing on cleaning up a few pipeline details for another RC3 release in the next week or 2. so it will be in there. For now, this feature is only compatible with development master of Logstash

**Pull request merged**